### PR TITLE
BUG GH23451 Allow setting date in string index for Series

### LINKED ISSUE #23451: Can't put date in Series if index is a string longer than 1 character
#### Code Sample

```
>>> import pandas
>>> x = pandas.Series([1,2,3],  index=['Date','b','other'])
>>> x
Date     1
b        2
other    3
dtype: int64
>>> from datetime import date
>>> x.Date = date.today()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python37\lib\site-packages\pandas\core\generic.py", line 4405, in __setattr__
    self[name] = value
  File "C:\Python37\lib\site-packages\pandas\core\series.py", line 939, in __setitem__
    setitem(key, value)
  File "C:\Python37\lib\site-packages\pandas\core\series.py", line 935, in setitem
    self._set_with(key, value)
  File "C:\Python37\lib\site-packages\pandas\core\series.py", line 983, in _set_with
    self._set_labels(key, value)
  File "C:\Python37\lib\site-packages\pandas\core\series.py", line 993, in _set_labels
    raise ValueError('%s not contained in the index' % str(key[mask]))
ValueError: ['D' 'a' 't' 'e'] not contained in the index
>>> x.b = date.today()
>>> x.b
datetime.date(2018, 11, 1)
>>> x
Date              1
b        2018-11-01
other             3
dtype: object
>>>
```
#### Problem description

I cannot put a date object in a Series if the index is a string with len > 1.
It works if it's only a single character.  Other types seem to work.  I've only seen the problem with dates.

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #23451 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Prevents string indices from turning into lists of characters when setting a date.

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -947,6 +947,9 @@ def _set_with(self, key, value):
                 except Exception:
                     pass
 
+            if isinstance(key, str):
+                key = [key]
+
             if not isinstance(key, (list, Series, np.ndarray, Series)):
```
Comment (jreback) [inline pandas/core/series.py]: Series is here twice :< can you remove

Comment (jreback) [reply pandas/core/series.py]: this can be an elseif

File: pandas/core/series.py [hunk 2]
```diff
@@ -947,6 +947,9 @@ def _set_with(self, key, value):
                 except Exception:
                     pass
 
+            if isinstance(key, str):
```
Comment (jreback) [inline pandas/core/series.py]: you might be able to make this is_scalar

File: pandas/tests/series/test_datetime_values.py [hunk 3]
```diff
@@ -548,3 +548,9 @@ def test_minmax_nat_series(self, nat):
     def test_minmax_nat_dataframe(self, nat):
         assert nat.min()[0] is pd.NaT
         assert nat.max()[0] is pd.NaT
+
+    def test_set_dt_with_string_index(self):
+        from datetime import date
```
Comment (jreback) [inline pandas/tests/series/test_datetime_values.py]: imports go at the top

File: pandas/tests/series/test_datetime_values.py [hunk 4]
```diff
@@ -548,3 +548,9 @@ def test_minmax_nat_series(self, nat):
...
+        x = pd.Series([1, 2, 3], index=['Date', 'b', 'other'])
+        x.Date = date.today()
```
Comment (jreback) [inline pandas/tests/series/test_datetime_values.py]: can you try this both as (parametrize)
``x.Date = `` and ``x['Date'] = ``

**Review by jreback — CHANGES_REQUESTED** [review]
needs a whatsnew note as well

**Comment by mroeschke** [comment]
Can you see if this fixes https://github.com/pandas-dev/pandas/issues/12862 as well?

File: pandas/tests/series/test_datetime_values.py (hunk 4 above)
Comment (JustinZhengBC) [reply pandas/tests/series/test_datetime_values.py]: I can't seem to figure out how to parametrize both these cases, as "Date" is being used in different contexts. What would be the variable(s) I parametrize and how would I use it with `x` to cover both cases?

**Comment by JustinZhengBC** [comment]
@mroeschke it fixes the issue of strings being split into lists, but that bug still persists because for tz-aware timestamps, the data is stored in a `DateTimeIndex`, whereas almost every other data format is stored in an `ndarray`, and setting an item directly requires an `ndarray` to be passed to the [backend function](https://github.com/pandas-dev/pandas/blob/24ab22f75aae1c9b9491c26e6f753814b7d6fd8e/pandas/_libs/index.pyx#L95). After this, I can do a quick PR to make the [] operator work just like using `.loc`, but it would still lose the timezone data after adding a new item. @jreback is that okay or should the problem be fixed closer to the root cause?

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -1115,6 +1115,7 @@ Datetimelike
 - Bug in :func:`DataFrame.combine` with datetimelike values raising a TypeError (:issue:`23079`)
 - Bug in :func:`date_range` with frequency of ``Day`` or higher where dates sufficiently far in the future could wrap around to the past instead of raising ``OutOfBoundsDatetime`` (:issue:`14187`)
 - Bug in :class:`PeriodIndex` with attribute ``freq.n`` greater than 1 where adding a :class:`DateOffset` object would return incorrect results (:issue:`23215`)
+- Bug in :class:`Series` that interpreted string indices as lists of characters when setting datetimelike values
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add the issue number

File: pandas/tests/series/test_datetime_values.py [hunk 6]
```diff
@@ -548,3 +548,10 @@ def test_minmax_nat_series(self, nat):
     def test_minmax_nat_dataframe(self, nat):
         assert nat.min()[0] is pd.NaT
         assert nat.max()[0] is pd.NaT
+
+    def test_set_dt_with_string_index(self):
```
Comment (jreback) [inline pandas/tests/series/test_datetime_values.py]: rename to test_setitem_with_string_index

**Review by jreback — CHANGES_REQUESTED** [review]
small comments, ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by JustinZhengBC** [comment]
@jreback green now

**Comment by jreback** [comment]
thanks!

**Pull request merged**